src/index.ts

```diff
@@ -284,377 +201,289 @@ export interface TOTPStrategyOptions {
-            const expiresAt = new Date(expiresAtEpochMs)
-
-            await this.createTOTP(
-              { hash: signedTotp, active: true, attempts: 0 },
-              expiresAt,
-            )
-
-            // Send TOTP.
-            await this.sendTOTP({
-              email: formDataEmail,
-              code: _otp,
-              magicLink,
-              form: formData,
-              request,
-            })
-
-            session.set(this.sessionEmailKey, formDataEmail)
-            session.set(this.sessionTotpKey, signedTotp)
-            session.set(this.sessionTotpExpiresAtKey, expiresAt.toISOString())
-            session.unset(options.sessionErrorKey)
-
-            throw redirect(options.successRedirect, {
-              headers: {
-                'set-cookie': await sessionStorage.commitSession(session, {
-                  maxAge: this.maxAge,
-                }),
-              },
-            })
-          }
-        }
-
-        /**
-         * 2nd Authentication Phase.
-         * Either via form submission or magic-link URL.
-         */
-        if (isGET && this.magicLinkGeneration.enabled) {
-          const url = new URL(request.url)
-
-          if (url.pathname !== this.magicLinkGeneration.callbackPath) {
-            throw new Error(ERRORS.INVALID_MAGIC_LINK_PATH)
-          }
-
-          magicLinkTotp = url.searchParams.has(this.totpFieldKey)
-            ? decodeURIComponent(url.searchParams.get(this.totpFieldKey) ?? '')
-            : undefined
-        }
-
-        if ((isPOST && formDataTotp) || (isGET && magicLinkTotp)) {
-          // Validation.
-          if (!sessionEmail || !sessionTotp || !sessionTotpExpiresAt) {
-            throw new Error(this.customErrors.inactiveTotp)
-          }
-
-          const expiresAt = new Date(sessionTotpExpiresAt)
-
-          if (isPOST && formDataTotp) {
-            await this._validateTOTP(sessionTotp, formDataTotp, expiresAt)
-          }
-          if (isGET && magicLinkTotp) {
-            await this._validateTOTP(sessionTotp, magicLinkTotp, expiresAt)
-          }
-
-          // Invalidation.
-          await this.updateTOTP(sessionTotp, { active: false }, expiresAt)
-
-          // Allow developer to handle user validation.
-          user = await this.verify({
-            email: sessionEmail,
-            form: formData,
-            magicLink: magicLinkTotp,
-            request,
-          })
-
-          session.set(options.sessionKey, user)
-          session.unset(this.sessionEmailKey)
-          session.unset(this.sessionTotpKey)
-          session.unset(this.sessionTotpExpiresAtKey)
-          session.unset(options.sessionErrorKey)
-
-          throw redirect(options.successRedirect, {
-            headers: {
-              'set-cookie': await sessionStorage.commitSession(session, {
-                maxAge: this.maxAge,
-              }),
-            },
-          })
-        }
+      if (email) {
+        await this._generateAndSendTOTP({
```

Comment:
Not a big fan of this function that does all those things, although I get the point and it looks clean, I prefer having each function do one thing. This also sets the session and that's not explicit in the name of the function.

Could we refactor this as it was before? Or at least have separate functions for each task? E.g., one that generates the TOTP and signs it, another that sends the email, and another that sets the session (if required); if not, we can set the session without functions.

Although it looks clean on paper, it also hides some complexity. It reminds me of the `handleTOTP` we just deleted to get to v2.

Reply:
Refactored.